### DISCUSSION TITLE
Coordinating extension behavior between local and remote extension hosts

### DISCUSSION THREAD

**Discussion by davidwallacejackson** [microsoft/vscode-discussions#143] [Extension Development QnA] [answered]
I'm looking to create an extension that runs simultaneously in the local and the remote extension hosts when using dev containers, with the two sides working in cooperation. I know that [openExternal](https://code.visualstudio.com/api/advanced-topics/remote-extensions#opening-something-in-a-local-browser-or-application) can start a process on the local machine if I want it to, but I really want the extension to manage it — the idea being that the extension acts as a bridge to/from the containerized remote host, similar to what the Remote Containers extension itself is doing.

For better context: I’m trying to hack up a way for https://www.cypress.io/ to work in a dev container. The simplest way of doing this is to [literally run the test browser in Linux inside the container and forward a VNC connection], (https://devopstar.com/2022/01/03/cypress-testing-in-devcontainers-and-github-codespaces) but in practice, I’ve found that the performance isn’t great — particularly when the app has to make a lot of network requests. For some reason, the app runs much better if the browser is running on the host OS, even if the requests are ultimately being handled inside the container.

So I’m looking for a way to run the tests inside the container and externalize only the browser. In theory this should be doable (Cypress controls e.g. Chrome over TCP with the Chrome DevTools Protocol, and there’s no inherent reason that connection needs to be to a browser on the same machine), but it’ll require me to launch a browser on the outside and potentially run some kind of tunnel back into the container (as far as I know, the dev container port forwarding mechanism only works one way).

I’ve never made an extension before, but it seems like the best thing to do here is to design the extension as two services that work together (one on the remote side and one on the local side). I don’t want a fire-and-forget sort of a thing, and I don’t want users to have to install separate software on the host. There’s gotta be a way to do this with the extension API, right?

Tagging @isidorn, who responded to me in Slack and asked me to open a thread here.

**Comment by isidorn** [comment, marked as answer]
Thanks for opening this discussion. Let's leave give some time for the community to answer, and if nobody helps I can ask some devs that work on containers in our team.

**Reply by davidwallacejackson** [reply]
@isidorn, following up -- would it make sense to refer this to the team now?

**Reply by isidorn** [reply]
Maybe @chrmarti could help :)

**Reply by chrmarti** [reply]
The same extension can run only once per VS Code window either locally or remotely. You could have two extensions, but there is no direct way for them to communicate.

There is port forwarding where the host browser could connect to a local port that is then forward into the container. Is that what you need or would the container side have to initiate the connection?

**Reply by davidwallacejackson** [reply]
@chmarti For this to work in the way I'm envisioning, the container would have to call out to a server run by the client machine -- I know that's a bit weird. The arrangement I'm picturing is:

* A node service (Cypress) runs in the container
* My extension starts a new browser on the client machine with devtools enabled (normally, Cypress would do this itself on the same machine where it's running)
* My extension opens a tunnel allowing Cypress inside the container to hit the devtools of the browser outside

There are two alternatives I've found, and neither one is great:

1. [run a desktop in the container](https://devopstar.com/2022/01/03/cypress-testing-in-devcontainers-and-github-codespaces) (this is very janky and very slow)
2. bind-mount out of the container so that the client machine can run Cypress in its own OS (very janky in its own way, and also doesn't work with codespaces)

I think there's a fundamental question here that vscode + remote dev extension need to answer, which is, "what is the best practice for iterating on UI tests in a devcontainer-based environment?"

One possible answer is that Cypress should redesign itself to run in a standard, unmodified browser without elevated permissions, but this seems a bit restrictive. Cypress can drive Electron apps, for instance, and I don't know how those tests would work under that restriction. It also means that Cypress would have to completely remove its dependency on the devtools protocol.

Cypress and browser integration tests are what I'm most familiar with, but I suspect that there are other cases where this functionality (the ability to mandate that something must run locally while being able to communicate remotely) would be desirable -- desktop application development generally, mobile development, or hardware projects.

**Reply by chrmarti** [reply]
Not sure why running the browser inside the container doesn't work well. Maybe investigate that and try allocating more CPUs and memory for Docker (Docker Desktop has a settings UI for this). Also give X11 forwarding (instead of VNC) a try if you haven't already.

**Reply by davidwallacejackson** [reply]
CPUs are already at max, but I can try increasing the memory allocation -- OOM might explain the performance I saw. I did also try X11 forwarding (or rather, a browser-based X11 client), which still had some problems with getting out of sync with the state of the remote desktop and requiring a restart. I'm currently externalizing Cypress entirely (per my option 2) and that is working well, with the caveats I mentioned above.

Do you feel strongly that the desktop-in-container approach is the right one for these cases, going forward?

**Reply by chrmarti** [reply]
I'm not sure if desktop-in-container is the best approach, it just seemed simpler than having two extensions - one on the UI, one on the workspace side - working together. Maybe the option you're trying now is better/easier?

**Reply by davidwallacejackson** [reply]
What I'm doing now is working better than desktop-in-container, though it definitely was harder to get set up and (IMO) comes at a significant cost in terms of requiring dependencies on the local machine. I did realize that if I went the extension route I could bundle them as an extension pack and that would probably make the install experience less weird. But on further reflection I'm not sure I want to depend on vsc functionality here -- I'm trying to get us to a point where my colleagues who prefer jetbrains also have the option to use that in our devcontainer. And nothing about this is inherently editor-related; it's more that an extension is the most obvious way I've seen so far to package this functionality. I feel like there's something else that can be done here, but I'm not sure what. Still thinking about it.
